xdsl/transforms/loop_hoist_memref.py

```diff
@@ -0,0 +1,144 @@
+from dataclasses import dataclass
+
+from xdsl.dialects import builtin, memref, scf
+from xdsl.ir import Block, Operation, Region
+from xdsl.irdl import Operand
+from xdsl.passes import MLContext, ModulePass
+from xdsl.pattern_rewriter import (
+    GreedyRewritePatternApplier,
+    PatternRewriter,
+    PatternRewriteWalker,
+    RewritePattern,
+    op_type_rewrite_pattern,
+)
+from xdsl.transforms.utils import (
+    find_same_target_store,
+    get_operation_at_index,
+    is_loop_dependent,
+)
+
+
+@dataclass
+class LoopHoistMemref(RewritePattern):
+    """Hoist pairs of memref.loads and memref.stores out of a loop."""
```

Comment:
Trying to add some context to the rewrite as an excuse to add some familiar words (e.g.: redundant):

```suggestion
    """Hoist pairs of memref.loads and memref.stores out of a loop.
    
    This rewrite hoist pairs of redundant memref.load and memref.store operations outside of their enclosing scf loop nest. The memref operation pair is considered for this rewrite if their memref target location is the same and it is constant with respect to the containing loop's induction variable.
    """
```

Reply:
`redundant` is a bit vague, redundant stuff can get DCE'd in my mind. But I'll add some more "words"